graph-builder/src/registry.rs

```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+type RegistryClient = dkregistry::v2::Client;
+
+fn registry_trim(src: &str) -> &str {
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(
+    credentials_file: &str,
+) -> Result<Box<dyn std::io::Read>, Error> {
+    if credentials_file != "" {
```

Comment:
Use an `Option` instead of special-casing an empty string. This entire function can be reduced to:

```rust
  Ok(Box::new(credentials_path.map_or(std::io::empty(), |path| File::open(path)?)))
```

It might be easier to get rid of this function completely since the above expression is so simple. It would also avoid the boxing.

Reply:
If I write the function as suggested:

```rust
fn get_credentials_or_empty_reader(
    credentials_path: &Path,
) -> Result<Box<dyn std::io::Read>, Error> {
    Ok(Box::new(
        credentials_path.map_or(std::io::empty(), |path| File::open(path)?),
    ))
}
```

the compiler errors out:


```
   Compiling graph-builder v0.1.0 (/home/steveej/src/github/openshift/cincinnati/graph-builder)                                              
error[E0599]: no method named `map_or` found for type `&std::path::Path` in the current scope                                                
  --> graph-builder/src/registry.rs:58:26                                                                                                    
   |                                                                                                                                         
58 |         credentials_path.map_or(std::io::empty(), |path| File::open(path)?),                                                            
   |                          ^^^^^^                                                                                                         
                                                                                                                                             
error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
  --> graph-builder/src/registry.rs:58:58                                                                                                    
   |                                                                                                                                         
58 |         credentials_path.map_or(std::io::empty(), |path| File::open(path)?),                                                            
   |                                                          ^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `std::fs::File`
   |                                                                                                                                         
   = help: the trait `std::ops::Try` is not implemented for `std::fs::File`                                                                  
   = note: required by `std::ops::Try::from_error`                                                                                           
                                                                                                                                             
error[E0277]: the size for values of type `dyn std::io::Read` cannot be known at compilation time                                            
  --> graph-builder/src/registry.rs:57:8                                                                                                     
   |                                                                                                                                         
57 |     Ok(Box::new(                                                                                                                        
   |        ^^^^^^^^ doesn't have a size known at compile-time                                                                               
   |                                                                                                                                         
   = help: the trait `std::marker::Sized` is not implemented for `dyn std::io::Read`                                                         
   = note: to learn more, visit <https://doc.rust-lang.org/book/second-edition/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
   = note: required by `<std::boxed::Box<T>>::new`                                                                                           
                                                                                                                                             
error: aborting due to 3 previous errors                                                                                                     
                                                                                                                                             
Some errors occurred: E0277, E0599.                                                                                                          
For more information about an error, try `rustc --explain E0277`.                                                                            
error: Could not compile `graph-builder`.                                                                                                    

To learn more, run the command again with --verbose.
error: Recipe `run-graph-builder` failed on line 15 with exit code 101
```

If your motivation is simplicity then I suggest we leave it as is. I initially tried to use `Path` for this and ran into complications that didn't seem worth the effort. I'll take another closer look tomorrow.

Reply:
I went for `Option<PathBuf>` since that's [officially supported by structopt](https://docs.rs/structopt/0.2.12/structopt/?search=#how-to-derivestructopt) and was the least effort. I couldn't get around using a separate function returning a the trait object for declaring the implementation of `std::io::Read`.